### ISSUE TITLE
Intermittent r10k hangs

### ISSUE THREAD

**Issue by esalberg** [puppetlabs/r10k#409]
After upgrading to r10k 1.5.0 (via commit 584529c3df of the r10k module, which was the latest as of a couple of days ago), we are seeing an intermittent hang of git during the postrun of r10k.  The hang does not appear to be related to a specific module that is being fetched.

The postrun command is basic:
r10k deploy environment -p

An example from the process list:

```
root     28355 22798  0 Apr21 ?        00:00:18 puppet agent: applying configuration
root     29477 28355  0 Apr21 ?        00:00:32 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p
root      4975 29477  0 Apr21 ?        00:00:00 git --git-dir /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-java.git fetch --prune
root      4978  4975  0 Apr21 ?        00:00:00 git-remote-https origin https://github.com/puppetlabs/puppetlabs-java.git
root      4985  4978  0 Apr21 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --include-tag --thin --no-progress https://github.com/puppetlabs/puppetlabs-java.git  efs/pull/109/head efs/pull/109/merge
```

From strace, the r10k process is waiting and the git processes are reading.

Weirdly, the agent reports successful runs through 06:11 am on Apr 21 in the console and 07:12 am on Apr 21 in /var/log/messsages before the run reports / finished catalog run lines stop (despite the hanging procs).  There also isn't a run reported in the console at 00:00 (despite the time stamp of the hanging procs) - the runs on this node were occurring at around 11 and 41 minutes each hour.

I had reported a similar issue with PE 3.3.2 (much older version of r10k) and webhook runs - the issue appears to have improved greatly with upgrades and a switch to deploy keys for our local code.  However, the issue related to the postrun command hanging the Puppet agent run appears to be new (it puts my masters into the Unresponsive list in the console, which I don't remember happening before).  I have a separate ticket open with Puppet support regarding the general supposition that the agent should not hang even if the postrun command never completes.

I am able to run a successful (no errors) "r10k deploy environment -pv" from the command line even while these processes are hanging.  My guess is that the connection to Github was slow enough (it can sometimes seem laggy in the browser, etc.) or otherwise momentarily problematic and disrupted the read, and that's why a later run works fine, but I don't know if there are any other differences regarding Puppet kicking off an r10k deploy versus my running it ad hoc from the command line.

I'm attempting to work around the issue by using tag and commit refs for as many modules as possible.

Is there anything that I can do to try and troubleshoot this further?  I'm more than happy to gather information, but I haven't been able to reproduce this issue on command yet.  

 (I'd be fine with some setting to end git calls that had not returned after a certain period of time / returned an error, just so that my server wouldn't be stuck without notice.)

**Comment by adrienthebo** [comment]
First off, you should upgrade to r10k 1.5.1 - there was a regression in 1.5.0 that cause Git modules to not correctly track branch updates.

Are you able to determine which file descriptor the read is hanging on? I'm guessing it's a network socket but it would be nice to confirm that.

In addition r10k 1.5.x provides support for using libgit2 instead of shelling out to Git; https://github.com/puppetlabs/r10k/blob/master/doc/git/providers.mkd#rugged documents the new behavior and settings. You can try using the rugged provider to see if that sidesteps the issue. As an added benefit it's faster than shelling out to Git.

If using the rugged provider fails then I'm guessing that you're running into https://tickets.puppetlabs.com/browse/RK-85 or some variation on that; I can bump the priority of that issue and maybe get it fixed in 1.5.2, or 1.6.0 at the latest.

**Comment by esalberg** [comment]
Yes!  That bug looks exactly like it!  If you can increase the priority, that would be great, but I will see what I can do to workaround the issue.

[ETA:] I double-checked and I am actually running r10k 1.5.1, so that's all set at least.  I'll try setting up the rugged provider as well with the docs you linked.

Looking at the strace output:

```
[pid 4975] read(4, 
```

 --> ls output - the pipe links are broken

```
[root@server ~]# ls -l /proc/4975/fd
lr-x------ 1 root root 64 Apr 22 12:50 0 -> /dev/null
l-wx------ 1 root root 64 Apr 22 12:50 1 -> pipe:[53650892]
l-wx------ 1 root root 64 Apr 22 12:50 2 -> pipe:[53650893]
lr-x------ 1 root root 64 Apr 22 12:50 3 -> /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-java.git/objects/pack/pack-a2caa73d818613cca92ac8de15ba5b5d60fd3aec.pack
lr-x------ 1 root root 64 Apr 22 12:50 4 -> pipe:[53650897]
l-wx------ 1 root root 64 Apr 22 12:50 5 -> pipe:[53650896]
lr-x------ 1 root root 64 Apr 22 12:50 6 -> pipe:[53650897]
```

 --> lsof output:

```
[root@server ~]# lsof -i | grep 4975
(nothing)
```

```
[pid 4978] read(8,
```

 --> ls output - the pipe and socket links are all broken

```
[root@linvm2153-2 ~]# ls -l /proc/4978/fd
lr-x------ 1 root root 64 Apr 22 12:48 0 -> pipe:[53650896]
l-wx------ 1 root root 64 Apr 22 12:48 1 -> pipe:[53650897]
l-wx------ 1 root root 64 Apr 22 12:48 2 -> pipe:[53650893]
lrwx------ 1 root root 64 Apr 22 12:48 3 -> socket:[53650908]
lrwx------ 1 root root 64 Apr 22 12:48 4 -> /etc/pki/nssdb/cert9.db
lrwx------ 1 root root 64 Apr 22 12:48 5 -> /etc/pki/nssdb/key4.db
lrwx------ 1 root root 64 Apr 22 12:48 6 -> socket:[53650922]
l-wx------ 1 root root 64 Apr 22 12:48 7 -> pipe:[53650916]
lr-x------ 1 root root 64 Apr 22 12:48 8 -> pipe:[53650917]
```

 --> lsof output:

```
[root@server ~]# lsof -i | grep 4978
git-remot  4978        root    3u  IPv4 53650908      0t0  TCP server.company.com:27191->github.com:https (CLOSE_WAIT)
git-remot  4978        root    6u  IPv4 53650922      0t0  TCP server.company.com:27192->github.com:https (CLOSE_WAIT)
```

```
[pid 4985] read(0,
```

 --> ls output - the pipe and socket links are all broken

```
[root@server ~]# ls -l /proc/4985/fd/
lr-x------ 1 root root 64 Apr 22 12:50 0 -> pipe:[53650916]
l-wx------ 1 root root 64 Apr 22 12:50 1 -> pipe:[53650917]
l-wx------ 1 root root 64 Apr 22 12:50 2 -> pipe:[53650893]
lrwx------ 1 root root 64 Apr 22 12:50 3 -> socket:[53650908]
lr-x------ 1 root root 64 Apr 22 12:50 4 -> /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-java.git/objects/pack/pack-a2caa73d818613cca92ac8de15ba5b5d60fd3aec.pack
```

 --> lsof output:

```
[root@server ~]# lsof -i | grep 4985
git        4985        root    3u  IPv4 53650908      0t0  TCP server.company.com:27191->github.com:https (CLOSE_WAIT)
```

**Comment by esalberg** [comment]
FYI - I have setup the rugged git provider (and am extending the r10k puppet module to handle it).  I'm pretty sure that it's in use, because I needed to install dependencies, and before I did that, r10k threw an error stating that "Git provider 'rugged' is not functional".

Is there anywhere I can look (logs, etc.) to validate that the rugged provider is in use / configured properly?

(Dependencies needed: cmake rpm, rugged gem)

**Comment by adrienthebo** [comment]
There's a commit in master (https://github.com/puppetlabs/r10k/commit/94bfeebc1e8f4a723187c60c78ab5fcde2e1bd39) that will log warnings if rugged wasn't built correctly, but it didn't make the cut for 1.5.0 or 1.5.1. I'll try to backport it into 1.5.2.

With respect to rugged, the gem ships with a vendored copy of libgit2 which has the unfortunate side effect of making it hard to make sure that it's getting built correctly. If you're able to install libgit2 via the system and make sure it's linked against libssh2 then you should be set, and then you can install rugged with `gem install rugged --use-system libraries`. The entire process of getting libgit2 compiled with support for libssh2 and all that can be a pain so PE 3.8 will ship with a copy of rugged/libgit2 precompiled with all of the necessary support, and since PE 3.8 should ship before the end of April you might be able to hold out for a while and avoid compiling it.

If you are eager to try to compile libgit2 right now I can see about walking you through the steps now, but there are a number of quirks and bugs in libgit2/libssh2 that make it a pain to get things compiled correctly so there might be a fair amount of debugging involved.

With respect to debugging the existing shellgit provider, it looks like the process with pid 4978 is blocked while reading on fd 8, which is pipe:[53650917]. If there was an issue with the connection to GitHub I would have figured it would be blocking on a read against one of the socket fds. If you're able to catch this deadlocked state, could you determine what's on the other end of that pipe? Doing something like `lsof | grep 53650917` should turn up the offending process. My best guess for this problem is to trace down the source of the hang via strace, /proc/<pid>, and lsof - once we know what everything is hanging on then we might be able to actually resolve it. Down the road I'll definitely add a timeout which will sidestep the problem but now I'm interested in actually figuring out what's causing the deadlock.

One thing that jumped out at me is that you're using an https URL for contacting GitHub; have you ever seen this deadlock occur for processes using the git protocol?

**Comment by esalberg** [comment]
I think I'm fine with waiting for PE 3.8 - it does sound like getting libgit2 into a desirable state can take some tweaking.  We definitely can manage a few more weeks in our current state.  If it becomes necessary for investigation / troubleshooting, then we can reconsider at that point.

I -think- that the hanging proceesses have generally mentioned Github (and thus https), but I'm not 100% sure.  I think that some of the improvement we gained in upgrading was due to setting up deploy keys for accessing our local control / hiera repos over ssh (partially because Gitlab can be slow over https).  Other than that, I am using https for all other modules, especially Github modules that we don't have forked - I was assuming that was the best way to get read-only access to those modules.  Is that incorrect?
(Does r10k keep an open connection to the control repo, or does it download the Puppetfile and then run through it locally?)

I had to kill the processes on the server for which I previously sent the fd info, but luckily, one of my production servers has hanging processes still.  The lsof output doesn't look too exceptionally helpful, but I will leave the server in this state as long as I can / until you tell me that we've gathered all the info.

```
root     22886 10105  0 Apr17 ?        00:00:12 puppet agent: applying configuration
root     24621 22886  0 Apr17 ?        00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p
root     24688 24621  0 Apr17 ?        00:00:00 git --git-dir /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git fetch --prune origin
root     24691 24688  0 Apr17 ?        00:00:00 git-remote-https origin https://github.com/puppetlabs/puppetlabs-concat.git
root     24703 24691  0 Apr17 ?        00:00:00 git fetch-pack --stateless-rpc --lock-pack --include-tag --thin --no-progress https://github.com/puppetlabs/puppetlabs-concat.git  efs/pull/298/head  efs/pull/298/merge
```

24688: fd 4

```
lr-x------ 1 root root 64 Apr 20 20:32 4 -> pipe:[76972546]
```

 --> lsof | grep 76972546

```
git       24688        root    4r     FIFO                0,8      0t0   76972546 pipe
git       24688        root    6r     FIFO                0,8      0t0   76972546 pipe
git-remot 24691        root    1w     FIFO                0,8      0t0   76972546 pipe
```

24691: fd 8

```
lr-x------ 1 root root 64 Apr 20 20:32 8 -> pipe:[76972574]
```

 --> lsof | grep 76972574

```
git-remot 24691        root    8r     FIFO                0,8      0t0   76972574 pipe
git       24703        root    1w     FIFO                0,8      0t0   76972574 pipe
```

24703: fd 0

```
lr-x------ 1 root root 64 Apr 20 20:32 0 -> pipe:[76972573]
```

 --> lsof | grep 76972573

```
git-remot 24691        root    7w     FIFO                0,8      0t0   76972573 pipe
git       24703        root    0r     FIFO                0,8      0t0   76972573 pipe
```

Just out of curiosity, I ran the following:
 --> lsof | grep -i github

```
git       24688        root  mem       REG              253,4    60684       1467 /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git/objects/pack/pack-e7218c693943eb3dd8891621acba4623396c85a3.idx
git       24688        root    3r      REG              253,4   477899       1466 /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git/objects/pack/pack-e7218c693943eb3dd8891621acba4623396c85a3.pack
git-remot 24691        root    3u     IPv4           76972557      0t0        TCP server.company.com:43207->github.com:https (CLOSE_WAIT)
git-remot 24691        root    6u     IPv4           76972579      0t0        TCP server.company.com:43209->github.com:https (CLOSE_WAIT)
git       24703        root  mem       REG              253,4   477899       1466 /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git/objects/pack/pack-e7218c693943eb3dd8891621acba4623396c85a3.pack
git       24703        root  mem       REG              253,4    60684       1467 /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git/objects/pack/pack-e7218c693943eb3dd8891621acba4623396c85a3.idx
git       24703        root    3u     IPv4           76972557      0t0        TCP server.company.com:43207->github.com:https (CLOSE_WAIT)
git       24703        root    4r      REG              253,4   477899       1466 /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git/objects/pack/pack-e7218c693943eb3dd8891621acba4623396c85a3.pack
```

**Comment by adrienthebo** [comment]
When communicating with GitHub you can use both 'https' as well as the custom 'git' protocol; the reason I was asking about this is I was wondering if this might be a bug in the git-remote-https transport process.

Looking that this output makes me believe that either `git-remote-https` (24691) or `git fetch-pack` (24703) is deadlocking; in this environment is 24691 blocked while reading from fd 8 and 24703 is also blocked while reading from fd 0? If this is the case then that's your issue - the transport process and the fetch-pack process are deadlocked on each other. I have no idea what would cause this, but it does make sense.

**Comment by esalberg** [comment]
Assuming that I'm reading the strace properly, I believe you are correct:
strace -p 24703

```
read(0,
```

strace -p 24691

```
read(8,
```

I guess I just defaulted to using https to Github - I can try the git protocol as well (for some reason, I had thought it needed ssh keys - sorry about that!).

I did a quick Google search on deadlocks with fetch-pack and found some notes about past issues.  We are running git 1.7.1-3.el6_4.1 on RHEL 6.5, which according to rpm was built on 26-02-2013.  At least one of the issues was patched in 02-2013, so I'm assuming the fix is newer than the version we're running (I'm still tracking down an updated changelog - it might have been added in 1.7.10.1?).

http://www.spinics.net/lists/git/msg199557.html
- http://git.661346.n2.nabble.com/Deadlock-during-remote-update-td7626077.html
- comment states available since git 1.8.3 (I wasn't sure if the numbering was the same on all platforms)
- mentioned in git 1.7.10.1 release notes?:
  https://github.com/git/git/blob/398dd4bd039680ba98497fbedffa415a43583c16/Documentation/RelNotes/1.7.10.1.txt

We appear to have git 1.8.4 and 1.9.4 available from RHN - would you suggest that I upgrade the masters to one of those?  Are there any drawbacks that you know of?

(Our local Gitlab server is running Gitlab 6.8.1, which appears to have Git 1.9.3 bundled in.)

**Comment by adrienthebo** [comment]
The bug that you found does sound like the issue that we're seeing; without spending some quality time with gdb it would be hard to verify - but I don't have any other ideas. The version of Git that you're running vs the version of Git that the issue was fixed in does indicate that this is what you're seeing. (/cc @acidprime - you were seeing r10k/Git hang periodically as well - could you see which version of Git you are running on the affected server?)

As for remediating this, one first step might be to switch from the https protocol to the git protocol. Since the git protocol doesn't require auth as you've seen it should be a really simple matter of switching out the URL schemes in your r10k.yaml and Puppetfiles. I'm not familiar with how the git protocol works but I assume it's not using curl or the affected code paths under the hood so that might be a good way to help confirm the issue as well as getting things back into a working state.

If switching to the git protocol doesn't work then upgrading one of your masters to Git > 1.7.10 would be another good step. It's up to you as to whether you try 1.8 or 1.9.

Let's try going with these steps and seeing if one of these approaches solves your issue.

(By the way, providing the strace/lsof/procfs output has been extremely, extremely helpful. It has made this much easier to debug this, so thanks very much for doing that!)

**Comment by esalberg** [comment]
Sorry about the delay getting back to this ticket.  I was pulled off briefly onto another project.

I also tried switching to the git protocol, and it doesn't seem to be working well, so I wanted to make sure I wasn't doing something silly.  The references to git urls in the r10k documentation seem to use https consistently (which is why I used that to begin), as do the github connection notes (either https or ssh) - is switching protocol supposed to be as simple as changing "https://" to "git://"?  When I try that, I get Connection Refused messages, so either something is wrong in my git configuration somewhere, or my company is blocking the git protocol ports.

Unless I'm missing something obvious, I'm just going to go ahead with updating the git version (hopefully by the end of this week).

You're very welcome about the traces - happy to help. :)

**Comment by esalberg** [comment]
Quick update - I'm testing out the RedHat Software Collection for git19.  I guess these are generally enabled on demand - I used the "permanent" workaround of adding a file for /etc/profile.d/git.sh, which basically just adds a few environment variables:

```
export PATH=/opt/rh/git19/root/usr/bin${PATH:+:${PATH}}
export MANPATH=/opt/rh/git19/root/usr/share/man:${MANPATH}
```

I checked as myself and as root, and they are both showing git version 1.9.4.  However, these are interactive shells - I wanted to doublecheck as to whether you think this will also work when r10k is running its git commands (not an interactive shell).  Will the profile.d file be sourced?  Is there any way to verify which git version r10k is using (whether a quick test such as "git --version" as root, or from a debug message)?

**Comment by adrienthebo** [comment]
The value of PATH is going to depend on how you're running it; if you're running r10k via cron then that'll be determined cron itself; if you're running r10k with the puppet `prerun_command` setting then the PATH should be the value of the `path` fact. Whether profile.d is sourced is again dependent on how you're running r10k, as that depends on if bash is used to execute the command and if so how bash was invoked. R10k doesn't have a ready way of showing what version of Git it's using unfortunately; one option to get more insight is to have r10k run via whatever mechanism you choose, `kill -STOP ${pid}`, then run `tr '\0' '\n' < /proc/${pid}/environ` to see what the environment of the running process is.

**Comment by adrienthebo** [comment]
It's been about a week since I've heard anything about this issue; were you able to resolve this issue by upgrading Git?

**Comment by adrienthebo** [comment]
It's been about a month since this issue has seen an update. I'll leave this issue open for a bit longer in case this is still an issue; if there's no response in about a week I'm going to assume this issue is resolved and close this issue.

**Comment by esalberg** [comment]
Hi,

So sorry about the delay.  I am still working with some of the items (upgrading to 3.8 currently; still testing with the upgraded git), but I think there are enough plausible options at this time.  I don't know if you want to close the ticket and have me update it later when I have finished more testing, or provide feedback elsewhere, or only update if I still need more assistance - any of these would be fine.

Thanks again for all your help!

**Comment by adrienthebo** [comment]
It'll be really easy to reopen this if there are still issues; just comment on this and I can reopen it. Good luck on your upgrades!
